### DISCUSSION TITLE
Is it possible to control when a shortcode is rendered?

### DISCUSSION THREAD

**Discussion by jtlandis** [quarto-dev/quarto-cli#8252] [Q&A]
### Description

Similar to how a filter may be controlled when it is run by either placing `-at: pre-render`  or placing it before or after `quarto` under the yml header, is it possible to control when a shortcode is placed in the document?

The motivation is that I recently found [this word count extention](https://github.com/andrewheiss/quarto-wordcount), but so far it only prints to the console. I was working on creating a short code to make these values available via the meta data based on [this stack overflow thread](https://stackoverflow.com/questions/76968652/quarto-make-a-shortcode-that-reads-content-from-the-document), but it seems that the values are always zero because this word count filters are run at `pre-render`.

**Comment by jtlandis** [comment]
Actually, from looking into this further, it does seem like the short code does in fact run after the word count, so maybe we can close this discussion 😅

**Comment by mcanouil** [comment]
Note that extensions behaviours are not really up to Quarto but up to the authors of such extensions.
Better to ask the authors.

**Comment by jimjam-slam** [comment]
> Similar to how a filter may be controlled when it is run by either placing -at: pre-render or placing it before or after quarto under the yml header...

Technically a Lua [pre- or post-render script](https://quarto.org/docs/projects/scripts.html) is not a filter—if you want to use it as one, you need to read a document in in the script and call `:walk` on it yourself to run a filter. Just wanted to let you know, as it wasn't an obvious difference to me either!

**Reply by cderv** [reply]
I believe what @jtlandis is referring to is the new 1.4 feature that allows to target when a lua filter is run more precisely: https://quarto.org/docs/prerelease/1.4/lua_changes.html#more-precise-targeting-of-ast-processing-phases

Sharing in case you have not seen it yet

**Reply by jimjam-slam** [reply]
Oooh, I haven't! Apologies, @jtlandis!

(Also pinging @andrewpbray here because we're working on a filter together at the moment and he might be keen to see it!)

**Reply by andrewpbray** [reply]
Oh this is glorious!

**Reply by allefeld** [reply]
Somewhat related, I can create a new topic if that's better: As far as I can tell, shortcodes are implemented as Lua filters. That means they can create AST elements, but not input document text. However, the `include` shortcode does include another file in the input document text – how does that work? I'm asking because I'd like to implement something similar myself.

**Reply by mcanouil** [reply]
Quarto is not limited to Lua.
You can take a look at the codebase: https://github.com/quarto-dev/quarto-cli/blob/0b11aa23efdba15cb534a74a872b91fd527e7637/src/core/handlers/base.ts#L350-L385

**Reply by allefeld** [reply]
Okay, question answered, thanks! The `include` shortcode is implemented differently, not as a Lua/AST filter.

But as far as I can tell this functionality is available to the user only within project scripts, not as a filter in a custom format? At least the only reference to TypeScript I could find in the documentation is on the "Project Scripts" page, and the "Creating Shortcodes" page only mentions Lua. And the "something similar" I mentioned should work also outside of projects.

**Reply by mcanouil** [reply]
Are you talking about the "include" shortcode? As far as I know "input" does not exist and it is the second time you are using it.

**Reply by allefeld** [reply]
Yes, sorry, `include`. Edited.

**Reply by cderv** [reply]
For now, extending Quarto behavior with shortcode is only available with Lua. There is not extension mechanism available for all the handlers that are in Typescript code base. This includes special inline handlers that are using Shortcode syntax (like `include`), but also computation engine or special block handler like diagram's one `mermaid` or `dot`. 

Related Feature Request 
* https://github.com/quarto-dev/quarto-cli/issues/4839

> At least the only reference to TypeScript I could find in the documentation is on the "Project Scripts" page

Project Scripts is a feature for Quarto Project that allows a user to provide some pre/post render scripts to be execute in the `quarto render` process. These scripts can be written in Typescript and evaluated in Quarto Project context. 

This is different that extending Quarto behavior. Only what is described in our Extensions part of the doc is possible. This is with Lua  for now as part of what Pandoc allows to do.
